docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+
+Call this the `NetworkName`. It represents both the application being run and the community or intention
+of running it.
+
+Peers only connect to other peers with the same NetworkName.
+
+### Forks
+
+We need to support existing networks upgrading and forking, wherein they may do any of:
+	- revert back to some height, continue with the same versions but new blocks
+	- arbitrarily mutate state at some height, continue with the same versions (eg. Dao Fork)
+	- change the AppVersion at some height
+
+Note because of Tendermint's voting power threshold rules, a chain can only be extended under the "original" rules and under the new rules
+if 1/3 or more is double signing, which is expressly prohibited, and is supposed to result in their punishment on both chains. Since they can censor
+the punishment, the chain is expected to be hardforked to remove the validators. Thus, if both branches are to continue after a fork,
+they will each require a new identifier, and the old chain identifier will be retired (ie. only useful for syncing history, not for new blocks)..
+
+We need a consistent way to describe forks.
+
+## Ideal
+
+Each component of the software is independently versioned in a modular way and its easy to mix and match and upgrade.
+
+Good luck pal ;)
+
+## Proposal
+
+### Versions
+
+Each of BlockVersion, AppVersion, P2PVersion is a monotonically increasing int64.
+
+An alternative might be to try to use Amino - ie. include the version number in the registration name for some type,
+for instance the BlockVersion in the header. However this would require the registered-name of the header to change even
+if it wasn't the header type itself that changed, which could be confusing. Thus, versions here will be integers in Amino encoded
+structus. That said, Amino's flexibility (ie. both in registration of new types and proto3 extensibility in structs) will make
+actual upgrades smoother.
+
+#### BlockVersion
+
+BlockVersion is the first field in the block Header. Since we have settled on a proto3 header,
+the ability to read the BlockVersion out of the serialized header is unanimous.
+
+BlockVersion is included in the NodeInfo and we only connect to peers with the same or higher BlockVersion.
+
+In the event that we need to upgrade the core protocols and don't want to dump the state,
+we can condition execution on the block version and use different handlers / message types as needed.
+In such a case we may need to support connecting to old versions as well.
+
+This isn't nice, but should happen quite infrequently and ideally only for extreme emergency.
+
+Note Ethereum has not had to make an upgrade like this (everything has been at state machine level, AFAIK).
+
+#### P2PVersion
+
+P2PVersion is not included in the block Header.
+
+P2PVersion is the first field in the NodeInfo. NodeInfo is also proto3 so this is easy to read out.
+
+For each P2PVersion, we keep a list of previous versions that this one is compatible with so we can connect to
```

Comment:
Can we use a semantic version?

Reply:
Semantic version doesn't cut it - we want to be able to make a sequence of non-breaking changes that result in a breaking change. ie. v20 is compatible with v19, and v21 is compatible with v20, but v21 is not compatible with v19

Reply:
Agreed for not using semver, but for a different reason.

There may be multiple types of p2p broadcasts you want to support that are entirely distinct. I think we should list the supported p2p broacst methods, and both nodes list all the things that they support. It may be the case that eventually people want to support fancier broadcast methods. I recall an issue where zarko and jeremiah were discussing it, however I can't find that issue rn.



Maybe there should be a secret connection negotiation step, and a tx broadcast method (After authentication) negotiation step.